src/ray/object_manager/object_manager.cc

```diff
@@ -193,6 +203,14 @@ ray::Status ObjectManager::PullEstablishConnection(const ObjectID &object_id,
 
 ray::Status ObjectManager::PullSendRequest(const ObjectID &object_id,
                                            std::shared_ptr<SenderConnection> conn) {
+  if (ObjectInTransitOrLocal(object_id)) {
+    // Do nothing if the object is already being received.
+    // We check here too just in case the object ends up in transit
+    // in the time between this method call and the Pull method call.
+    RAY_CHECK_OK(
```

Comment:
`PullEstablishConnection` does not always directly invoke this method. It sometimes invokes an asynchronous call to `ObjectDirectory`.

Reply:
Why is it necessary for the call to `ObjectDirectory` to be asynchronous? Also, judging from the current implementation, it's actually a synchronous call since the `ObjectDirectory` calls the callbacks directly.

Also, I would rather we just did the check in this method then, to avoid grabbing the lock in the other methods.

Reply:
This should be okay now since we've removed the lock.